### DISCUSSION TITLE
Code Review for TypeBox Adapter for client-side form validation library

### DISCUSSION THREAD

**Discussion by lifeiscontent** [sinclairzx81/typebox#711] [Q&A] [answered]
Hi there, I've started to build out an adapter for TypeBox and a form validation library called conform, would love some feedback on the PR to see if there are some optimizations that could be done.

Have a look here: https://github.com/edmundhung/conform/pull/368

and thank you for your time!

**Comment by sinclairzx81** [comment]
@lifeiscontent Hi, 

The code looks fine, good work :) The only thing that stood out was maybe the use of `Static<T>` when decoding via `Value.Decode`. Note that the Decode function will return instances of `StaticDecode<T>` (so not `Static<T>`) with the main difference being that Decode will factor any Transform have applied to types (which may give a different inference type)

I've just quickly updated this code to infer as StaticDecode if you want to give this a test, but you will want to be mindful of pushing out TS inference (as StaticDecode can be a very expensive type to compute (as Transforms are mapped through the Type System)). The following is entirely optional, but if you want to use it, just try and gauge editor performance for large types first.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFAEgYE8wFM4DFhYA2AJgMIQB2AzjFAIbAUwA0qG2eBJVWM51tBk1ZpMOAMoBXAEYhgVKsEoiwdKDzh0qcaVqwAFNT1YBfOADMoEEHADkAAQDGlc9BABaGBAD0xa7YBuJFBIWDh2HAQ4cRg6GGBHZmjY+McAESxnYiwkgBUAeWkAK0yWOFzxRwALLBA6ODNLazt7RQpHQgYobwjpCAAPQODwaHgo-LB4yjpCAGlGYjyxAHFJNWIGiysbBzaOrp6xPsGgkNHEOAA1Gckcq5usAFEoKygASRgsei8oTaad1qMfbAbq9AbeABuDyGSHMknaUwo4Xo1FcUBAhhgVQAFKosQAuOA0KCMADmAEpCcSyYhUM4BHBVLBtABeRlxKoAOioYEIwBg2Ns3ls5M55mAhE+UGxACEIBBCFg6BRyQE4N5vNFefzwjUibQacqNlBahAIThapN0OzmagTTBJFAkUyYFRUChOXUwNjcWoygssP1yXAWQA+WkoSMauBvcy6nAuuDyTRwCiSEDSL5JNF1eA6rSaJFqejWgP9JIQLFfADu8hwkg09T88GgqcrVRpaPjFhBNBtMHdKHtjqR8gAcnQx9iANQu4MAfjg2LLIZZbIADHBF4nCQADTkAEgQLpMu+De4A2keTwBdXdBSMmcnuzlFCCMQUioImJBIQOheA4QRJQkVJXh8CIYgeD4ShiSEGAAB5cjgQNPgoKDykKEpHDKAAlFD+jQjDcMyaBiAQ6kKFJJJlXQcM2RiOIEgyLIsCQ0NQ2xVAqGqWo6EJXJWEpThIOg-g4MYRDcPDZAUCAnCQLgYh5EcdZZiwdBsShQhbkJEjnCgcj9CsbBYHQdT0Boig6KSABrDTCWMiBTIwCzg1klB6T7KIL3s9Ab0JAB9JJOVCk0+zMNltNuB8h14Ec4HCmAHx-OT4QUygkwocwvnEwRJKQgiiO0CpeLqTieJqOoBOEiCSDy+gCtyGT3UVeAvPyphCTqshYM6+A2QQExYuAONsVyFY1kMzk3ioCZERmbFKr48l3MHDq4Hhat6DAbANjZZSeLUjSlrKugknmkCZnmdDgwLUqqroWLPL6xqmBDCNIw9ULGByqAGvg7Etp2vbyREL6TQAR0kEEsGIQlzBmYxB2G90zCIDRRqXCbsFWdYZqoABBF46E05a6lWz6XoEN6Bqp77OV+3LXsB8m6EZz4QCoMHBxQEBJElYBeSwQlaFucGUFRx8UMITGxpxrA8em2aYhJKjTseymPJQLGNb4z1GAAGSwKisTWr7qYkpgDYoY3TaqD62Ztu3SSxZ7UsjXWnbqfoXbN+mOtpz06F9k3XYdtlvZDv2qndwcvbOzk8SlFUA5ZySk7iFPHcT5OvgoOPpYxnBdYVpXiAJsd00zaU2eDAAfevykm-HZreJgsDA2uzq1+Oxu9xhQHTc2vsD+CbZzx6baHkBC89-vE59meR8jMeM59yf9aX-nZ5Rvul29gX4mF-JzBXy3+u5T4wE3upPSPoXFVPuf0dlkv5Zb5WqAAVQoEDDf5F8RadcV5r2tnnJ0t92a0VPrzYO3psR8ilDMYMYZ6YXzgM5REH0kFAMIJyDqnIvCqzJNiVUvMUDRlyBAFCR0ODwioHQHKcBqhqDoDhL42hMomjAv0Chw5IFYJAhQj0JpeTsKwNibwF564AB1ZFDTIReWRN4AB6B5pwACp5ychvN4aidh5EHgAGQihEZyMRnRHCSO8O4fRSRbDyP6AAJmIF+XmT44Fvg-LYeu7jHy-nnkuCIEA4xgJgBYrA0NYb7TXHYeE2RxQUDhiKKm4TInRJNPtZEMU0aBLig6SB4SUqoHktg8KCpzQ9TEunJghVUImwwgUYopQkj4QaehbQ+kyIUQNFRKydEPqMVSCxCA2R2KcXdGzASIhapcCgrwAGBVpKfQEUiLCpRORgRgPkasFAnIuXQBOEAWAqB6zvmAEyXx4inNFFkyQ1ifTsMSHAPyqCWoQ3ipA7WDNnkSxQL5DSAUsp-SWUwLiFsUBO0uc5a5BAqCAutAWPyoSRLcEWbUxCwyEgTJvP8nmX0paSySENTQ2hqkYppvBBC0lyGS0CWsxKpzKlYApTBKlklzl0HIT+P8-QAIWHStgpkPAEKVEekVRpJUNk4TaUMlIzFSLjPFXxDiELVDoEIBAOg8M8BuDSHEeoTdv64UNuIJUUBqiGHoFzEQ9JxSkkJB5aZ0QzpwCbj6SSJsYBUj6RSEM4YVUUxSkJQkUhZDyEUJQGloYgh8oFWUxSIq2JBvqB0ppMq8LyqYukJVKazpqtQBqrVOrur6sNe6uAJqzUWqtWw21dIXDAEdZ9F1qbK2eo7l1fUat-VoNTXSkwobogyDkAoECMaIyMt0Dwa1orpK+k1dqxYqzmWEHNMuCgYBJD+i7TAUBfUiRupZLzEJYSm2km5MeuJthE2UFsBQxc9rm1XseputC+6KGEmfZetmz0NoVPXVgA1sQPpaQeGGhVjgxUFveUuKI0UcBPn-YewD5pnivDA18V4hJ4S2QoBAPZcGfm62w62RgNBlTWNRdcHSTwXjQA+EAn458CkJR+ZGMjUBCTExLGKbY2IuN3Lhg8yRgmGNQCSIhjD0BiMiIwXiCOyJlRUBzJiHE0mJOZzNs9C2G1GDJKgDJ34bJxOvAvIpm8W55xsgvDeOlkKdYUEM8ZpODYNMPFcychQdAwIOchYyrjunHwkrMAWbphlem9qSJRUkdnQwEotkSgJp6qhWGrChCT7s41fWjM4L41juyMG3fALw3Y2brUPcW5dH1aO3E5Pwc0sAuVJGKzu-ztBrQ-MZWh4DhrsR1awJyUZ2QWvsiXTq1a7sWFxGqEuIT9MetrvQxJgbDxOTGbOWzJI1XJv+Y9kO1AyGkAmCAA)

Nice work!
S

**Reply by lifeiscontent** [reply]
@sinclairzx81 thanks for the help! the author of conform wrote up a few comments on the PR about how to do a few different things in TypeBox that can be done in Zod/Yup, do you have a few minutes to take a look and uncover some of those questions? https://github.com/edmundhung/conform/pull/368

Thanks again for your help, and happy new year :)

**Reply by sinclairzx81** [reply, marked as answer]
@lifeiscontent Hiya,

> How do we add custom validation to typebox? For example, it's common to ask for the password twice in a signup form and validated whether both passwords match. How can we implement this using typebox?

A matching password check would need to be handled post validation

> Does typebox support async validation? If yes, can we also support it in the integration?

TypeBox does not support async validation, you will need to await values before checking them.

> How does the coercion work and how does it handle empty string?

Type coercion is supported via Value.Convert

---

I'm not sure what the requirements of conform are, but if it needs validation libraries to perform additional processing of values (above and beyond just checking the value), you can use additional operators from Value.* to construct a `parse` pipeline. These functions are typically used for processing values received on the server, but you can use them to parse values on the frontend also.

The following would be a "kitchen sink" setup (the `post` function is non-typical but added for the password check)

```typescript
// user defined parse function
function parse<T extends TSchema, R = StaticDecode<T>>(schema: T, value: unknown, post?: (value: R) => void): R  {
  const converted = Value.Convert(schema, value)     // apply type coercion if possible
  const defaulted = Value.Default(schema, converted) // apply defaults to missing properties
  const cleaned = Value.Clean(schema, defaulted)     // omit unknown properties
  const decoded = Value.Decode(schema, cleaned)      // decode the value 
  if(post) post(decoded as R)                        // (optional) post decode check
  return decoded as R
}
// user defined async parse function
async function parseAsync<T extends TSchema, R = StaticDecode<T>>(schema: T, promise: Promise<unknown>, post?: (value: R) => void) {
  return parse(schema, await promise, post)
}

// ---------

const T = Type.Object({
  password: Type.String(),
  confirm: Type.String()
})

const value = parse(T, {
  password: 1234567,
  confirm: 1234567
}, value => {
  if(value.password !== value.password) {
    throw Error('Passwords do not match')
  }
})

console.log(value)
```
You would likely want to embed the above internal to the plugin to process values on behalf of conform, however you'll have to contrast whether the above will work in the context of the library (as unlike Zod, TypeBox does not use combinators to process values). I got the sense that Zod-like combinators may be a "minimum barrier to entry" based on comments from the conform author, so you'll have to weigh whether the `post` validation check is tenable given the design of conform.

Hope this helps
S
